Source/ExecutionEngine/ImplementationTask.cs

```diff
@@ -62,40 +72,93 @@ public class ImplementationTask : IImplementationTask {
   }
 
   private CancellationTokenSource? cancellationSource;
+  private ReplaySubject<IVerificationStatus>? status;
 
   public void Cancel() {
     cancellationSource?.Cancel();
-    cancellationSource = null;
   }
 
+  public bool IsIdle => cancellationSource == null;
+
   public IObservable<IVerificationStatus>? TryRun()
   {
-    if (CacheStatus is Completed) {
+    lock (mayAccessCancellationSource) {
+      if (cancellationSource == null) {
+        return StartRunIfNeeded();
+      }
+
+      if (cancellationSource.IsCancellationRequested) {
+        // Another thread is running but was cancelled,
+        // so we may immediately start a new run after the cancellation completes.
+        return QueueRun();
+      }
+
+      // Another thread is running and is not cancelled, so this run fails.
       return null;
     }
```

Comment:
Does the whole method now run inside the lock?

Reply:
Jep